src/i2c.rs

```diff
@@ -0,0 +1,287 @@
+                    clocks: Clocks,
+                    apb: &mut APB1,
+                ) -> Self {
+                    apb.enr().modify(|_, w| w.$i2cXen().enabled());
+                    apb.rstr().modify(|_, w| w.$i2cXrst().set_bit());
+                    apb.rstr().modify(|_, w| w.$i2cXrst().clear_bit());
+
+                    let freq = mode.get_frequency();
+
+                    assert!(freq <= 400_000);
+
+                    let i2cclk = clocks.pclk1().0;
+                    let freqrange = (i2cclk / 1000000) as u16;
+
+                    i2c.cr2.modify(|_, w| unsafe {
+                        w.
+                            last().clear_bit().
+                            dmaen().clear_bit().
+                            itbufen().clear_bit().
+                            itevten().clear_bit().
+                            iterren().clear_bit().
+                            freq().bits(freqrange as u8)
+                    });
+                    i2c.cr1.write(|w| {
+                        w.
+                            pe().clear_bit()
+                    });
+
+                    match mode {
+                        Mode::Standard { .. } => {
+                            i2c.trise.write(|w| unsafe {
+                                w.trise().bits((freqrange + 1) as u8)
+                            });
+                            i2c.ccr.modify(|_, w| unsafe {
+                                w.ccr().bits(((i2cclk / (freq * 2)) as u16).max(4))
+                            });
+                        },
+                        Mode::Fast { duty_cycle, .. } => {
+                            i2c.trise.write(|w| unsafe {
+                                w.trise().bits((freqrange * 300 / 1000 + 1) as u8)
+                            });
+
+                            i2c.ccr.write(|w| {
+                                let (freq, duty) = match duty_cycle {
+                                    DutyCycle::Ratio1to1 => (((i2cclk / (freq * 3)) as u16).max(1), false),
+                                    DutyCycle::Ratio16to9 => (((i2cclk / (freq * 25)) as u16).max(1), true)
+                                };
+
+                                unsafe {
+                                    w.ccr().bits(freq).duty().bit(duty)
+                                }
+                            });
+                        }
+                    };
+
+                    i2c.cr1.modify(|_, w| w.pe().set_bit());
+
+                    I2c { i2c, pins }
+                }
+
+                /// Releases the I2C peripheral and associated pins
+                pub fn free(self) -> ($I2CX, PINS) {
+                    (self.i2c, self.pins)
+                }
+            }
+
+            impl<PINS> Write for I2c<$I2CX, PINS> {
+                type Error = Error;
+
+                fn write(&mut self, addr: u8, bytes: &[u8]) -> Result<(), Error> {
+                    // TODO support transfers of more than 255 bytes
```

Comment:
I think this TODO was left from the `stm32f30x-hal` implementation?

Reply:
Yep, but it's still valid

Reply:
Why do you say that? I didn't see anything suggesting this wouldn't work for larger transfers.